src/meta/src/controller/user.rs

```diff
@@ -0,0 +1,674 @@
+        let inner = self.inner.write().await;
+        let txn = inner.db.begin().await?;
+        check_user_name_duplicate(&pb_user.name, &txn).await?;
+
+        let grant_privileges = pb_user.grant_privileges.clone();
+        let user: user::ActiveModel = pb_user.into();
+        let user = user.insert(&txn).await?;
+
+        if !grant_privileges.is_empty() {
+            let mut privileges = vec![];
+            for gp in &grant_privileges {
+                let id = extract_grant_obj_id(gp.get_object()?);
+                for action_with_opt in &gp.action_with_opts {
+                    privileges.push(user_privilege::ActiveModel {
+                        user_id: Set(user.user_id),
+                        oid: Set(id),
+                        granted_by: Set(action_with_opt.granted_by),
+                        action: Set(action_with_opt.get_action()?.into()),
+                        with_grant_option: Set(action_with_opt.with_grant_option),
+                        ..Default::default()
+                    });
+                }
+            }
+            UserPrivilege::insert_many(privileges).exec(&txn).await?;
+        }
+        txn.commit().await?;
+
+        let mut user_info: PbUserInfo = user.into();
+        user_info.grant_privileges = grant_privileges;
+        let version = self
+            .notify_frontend(
+                NotificationOperation::Add,
+                NotificationInfo::User(user_info),
+            )
+            .await;
+
+        Ok(version)
+    }
+
+    async fn update_user(
+        &self,
+        update_user: PbUserInfo,
+        update_fields: &[PbUpdateField],
+    ) -> MetaResult<NotificationVersion> {
+        let inner = self.inner.write().await;
+        let rename_flag = update_fields
+            .iter()
+            .any(|&field| field == PbUpdateField::Rename);
+        if rename_flag {
+            check_user_name_duplicate(&update_user.name, &inner.db).await?;
+        }
+
+        let mut user = User::find_by_id(update_user.id)
+            .one(&inner.db)
+            .await?
+            .ok_or_else(|| MetaError::catalog_id_not_found("user", update_user.id))?;
+        update_fields.iter().for_each(|&field| match field {
+            PbUpdateField::Unspecified => unreachable!(),
+            PbUpdateField::Super => user.is_super = update_user.is_super,
+            PbUpdateField::Login => user.can_login = update_user.can_login,
+            PbUpdateField::CreateDb => user.can_create_db = update_user.can_create_db,
+            PbUpdateField::CreateUser => user.can_create_user = update_user.can_create_user,
+            PbUpdateField::AuthInfo => user.auth_info = update_user.auth_info.clone().map(AuthInfo),
+            PbUpdateField::Rename => user.name = update_user.name.clone(),
+        });
+
+        let user = user.into_active_model().update(&inner.db).await?;
+        let mut user_info: PbUserInfo = user.into();
```

Comment:
As the PK is also Set, I suspect the update will fail, according to my previous attempt.

Reply:
I remember that **Update** will update the columns that have been set according to the PK. Let me double check it.

Reply:
As PK is also Set in the ActiveModel, we'll get a duplicate key error, and that's why I need to handle onConflict in hummock models.

Reply:
Well, I think onConflict only works for `Insert`. I also added a unit test for `update_user` here and it works. https://github.com/risingwavelabs/risingwave/pull/13189/commits/1bde04ec0a1d77477f5c9eaefa245613c0ced6a0 . `SeaORM` requires each table must have a primary key, and I guess that's how `update` works, it will update all columns that have been `Set` according to the PK.

Besides, I also found an easily mistaken usage as I used before this commit. `into_active_model` will make all columns `Unchanged`, which will make update take no effects. But leave the PK `Unchanged` with original value also works here. It means whether changing it as `user.user_id = Set(update_user.id);` or not are both work.